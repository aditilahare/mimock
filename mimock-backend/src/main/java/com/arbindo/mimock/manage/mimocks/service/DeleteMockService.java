package com.arbindo.mimock.manage.mimocks.service;

public interface DeleteMockService {

    boolean hardDeleteMockById(String mockId);

    boolean softDeleteMockById(String mockId);

    boolean deleteAllMocks();

    void flushDeletedMocks();
}
