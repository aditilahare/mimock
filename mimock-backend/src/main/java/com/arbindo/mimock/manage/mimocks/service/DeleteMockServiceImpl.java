package com.arbindo.mimock.manage.mimocks.service;

import com.arbindo.mimock.common.services.EntityStatusService;
import com.arbindo.mimock.entities.EntityStatus;
import com.arbindo.mimock.entities.Mock;
import com.arbindo.mimock.repository.MocksRepository;
import com.arbindo.mimock.utils.ValidationUtil;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.extern.log4j.Log4j2;
import org.apache.logging.log4j.Level;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import javax.transaction.Transactional;
import java.time.ZonedDateTime;
import java.util.List;

@Service
@Log4j2
@Builder
@AllArgsConstructor
public class DeleteMockServiceImpl implements DeleteMockService {

    @Autowired
    private MocksRepository mocksRepository;

    @Autowired
    private EntityStatusService entityStatusService;

    @Autowired
    private MockManagementService mockManagementService;

    @Override
    public boolean hardDeleteMockById(String mockId) {
        if (ValidationUtil.isNotNullOrEmpty(mockId)) {
            try {
                Mock mock = mockManagementService.getMockById(mockId);
                if (mock != null) {
                    mocksRepository.delete(mock);
                    return true;
                }
            } catch (Exception e) {
                log.log(Level.DEBUG, e.getMessage());
                return false;
            }
        }
        log.log(Level.DEBUG, "Invalid Mock Id!");
        return false;
    }

    @Transactional
    @Override
    public boolean softDeleteMockById(String mockId) {
        if (ValidationUtil.isNotNullOrEmpty(mockId)) {
            try {
                Mock mock = mockManagementService.getMockById(mockId);
                if (mock != null) {
                    // Perform only soft delete i.e. Mark EntityStatus as DELETED
                    EntityStatus entityStatus = entityStatusService.getDeletedMockEntityStatus();
                    mock.setEntityStatus(entityStatus);
                    mock.setDeletedAt(ZonedDateTime.now());
                    mocksRepository.save(mock);
                    return true;
                }
            } catch (Exception e) {
                log.log(Level.DEBUG, e.getMessage());
                return false;
            }
        }
        log.log(Level.DEBUG, "Invalid Mock Id!");
        return false;
    }

    @Override
    public boolean deleteAllMocks() {
        try {
            mocksRepository.deleteAll();
            return true;
        } catch (Exception e) {
            log.log(Level.DEBUG, e.getMessage());
        }
        log.log(Level.DEBUG, "Unable to delete all mocks!");
        return false;
    }

    @Override
    public void flushDeletedMocks() {
        try {
            EntityStatus entityStatus = entityStatusService.getDeletedMockEntityStatus();
            ZonedDateTime thirtyDaysAgo = ZonedDateTime.now().minusDays(30);
            List<Mock> deletedMocks = mocksRepository.findAllByEntityStatusAndDeletedAt(entityStatus, thirtyDaysAgo);
            mocksRepository.deleteAll(deletedMocks);
            log.log(Level.INFO, "Flushed " + deletedMocks.size() + " mock(s)!");
            return;
        } catch (Exception e) {
            log.log(Level.DEBUG, e.getMessage());
        }
        log.log(Level.DEBUG, "Unable to flush deleted mocks!");
    }
}
