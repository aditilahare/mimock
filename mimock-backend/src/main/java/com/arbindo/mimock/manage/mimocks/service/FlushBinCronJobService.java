package com.arbindo.mimock.manage.mimocks.service;

public interface FlushBinCronJobService {
    void FlushBin();
}
