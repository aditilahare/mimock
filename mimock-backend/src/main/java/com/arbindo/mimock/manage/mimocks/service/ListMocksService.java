package com.arbindo.mimock.manage.mimocks.service;

import com.arbindo.mimock.entities.Mock;
import com.arbindo.mimock.manage.mimocks.enums.Status;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

public interface ListMocksService {

    List<Mock> getAllMocks();

    Page<Mock> getMocksAsPageable(Pageable pageable, Status status, String httpMethod, String expectedResponseType);
}
