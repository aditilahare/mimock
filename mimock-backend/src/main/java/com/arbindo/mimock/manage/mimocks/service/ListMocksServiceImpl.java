package com.arbindo.mimock.manage.mimocks.service;

import com.arbindo.mimock.common.services.EntityStatusService;
import com.arbindo.mimock.entities.EntityStatus;
import com.arbindo.mimock.entities.HttpMethod;
import com.arbindo.mimock.entities.Mock;
import com.arbindo.mimock.manage.mimocks.enums.ExpectedResponseType;
import com.arbindo.mimock.manage.mimocks.enums.Status;
import com.arbindo.mimock.manage.mimocks.service.helpers.MockParamBuilder;
import com.arbindo.mimock.repository.MocksRepository;
import com.arbindo.mimock.utils.ValidationUtil;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.extern.log4j.Log4j2;
import org.apache.logging.log4j.Level;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
@Log4j2
@Builder
@AllArgsConstructor
public class ListMocksServiceImpl implements ListMocksService {

    @Autowired
    private MocksRepository mocksRepository;

    @Autowired
    private EntityStatusService entityStatusService;

    @Autowired
    private MockParamBuilder mockParamBuilder;

    @Override
    public List<Mock> getAllMocks() {
        return mocksRepository.findAll();
    }

    @Override
    public Page<Mock> getMocksAsPageable(Pageable pageable, Status entityStatusParam,
                                         String httpMethodParam, String expectedResponseTypeParam) {
        HttpMethod httpMethod = null;
        EntityStatus entityStatus = null;
        ExpectedResponseType expectedResponseType = null;
        try{
            expectedResponseType = ExpectedResponseType.valueOf(expectedResponseTypeParam);
        }catch(Exception e){
            log.log(Level.DEBUG, e.getMessage());
        }
        if(ValidationUtil.isArgNotNull(httpMethodParam)){
            httpMethod = mockParamBuilder.findHttpMethodFromQueryString(httpMethodParam);
        }
        if (ValidationUtil.isArgNotNull(entityStatusParam)) {
            entityStatus = entityStatusService.findByEntityStatus(entityStatusParam.name());
        }

        if(ValidationUtil.isArgNotNull(httpMethod)
                && ValidationUtil.isArgNotNull(entityStatus)
                && ValidationUtil.isArgNotNull(expectedResponseType)){
            switch (expectedResponseType){
                case TEXTUAL_RESPONSE:
                    return mocksRepository.findAllByEntityStatusAndHttpMethodAndTextualResponseIsNotNull(entityStatus,
                            httpMethod, pageable);
                case BINARY_RESPONSE:
                    return mocksRepository.findAllByEntityStatusAndHttpMethodAndBinaryResponseIsNotNull(entityStatus,
                            httpMethod, pageable);
                case EMPTY_RESPONSE:
                    return mocksRepository.findAllByEntityStatusAndHttpMethodAndTextualResponseIsNullAndBinaryResponseIsNull(
                            entityStatus, httpMethod, pageable);
            }
        }

        if(ValidationUtil.isArgNotNull(httpMethod)
                && ValidationUtil.isArgNotNull(entityStatus)){
            return mocksRepository.findAllByEntityStatusAndHttpMethod(entityStatus, httpMethod, pageable);
        }

        if(ValidationUtil.isArgNotNull(entityStatus) && ValidationUtil.isArgNotNull(expectedResponseType)){
            switch (expectedResponseType){
                case TEXTUAL_RESPONSE:
                    return mocksRepository.findAllByEntityStatusAndTextualResponseIsNotNull(entityStatus, pageable);
                case BINARY_RESPONSE:
                    return mocksRepository.findAllByEntityStatusAndBinaryResponseIsNotNull(entityStatus, pageable);
                case EMPTY_RESPONSE:
                    return mocksRepository.findAllByEntityStatusAndTextualResponseIsNullAndBinaryResponseIsNull(entityStatus, pageable);
            }
        }

        if(ValidationUtil.isArgNotNull(httpMethod) && ValidationUtil.isArgNotNull(expectedResponseType)){
            switch (expectedResponseType){
                case TEXTUAL_RESPONSE:
                    return mocksRepository.findAllByHttpMethodAndTextualResponseIsNotNull(httpMethod, pageable);
                case BINARY_RESPONSE:
                    return mocksRepository.findAllByHttpMethodAndBinaryResponseIsNotNull(httpMethod, pageable);
                case EMPTY_RESPONSE:
                    return mocksRepository.findAllByHttpMethodAndTextualResponseIsNullAndBinaryResponseIsNull(httpMethod, pageable);
            }
        }

        if(ValidationUtil.isArgNotNull(httpMethod)){
            return mocksRepository.findAllByHttpMethod(httpMethod, pageable);
        }
        if (ValidationUtil.isArgNotNull(entityStatus)) {
            return mocksRepository.findAllByEntityStatus(entityStatus, pageable);
        }
        if(ValidationUtil.isArgNotNull(expectedResponseType)
                && expectedResponseType == ExpectedResponseType.TEXTUAL_RESPONSE){
            return mocksRepository.findAllByTextualResponseIsNotNull(pageable);
        }
        if(ValidationUtil.isArgNotNull(expectedResponseType)
                && expectedResponseType == ExpectedResponseType.BINARY_RESPONSE){
            return mocksRepository.findAllByBinaryResponseIsNotNull(pageable);
        }
        if(ValidationUtil.isArgNotNull(expectedResponseType)
                && expectedResponseType == ExpectedResponseType.EMPTY_RESPONSE){
            return mocksRepository.findAllByTextualResponseIsNullAndBinaryResponseIsNull(pageable);
        }
        return mocksRepository.findAll(pageable);
    }
}
