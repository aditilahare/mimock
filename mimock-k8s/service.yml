apiVersion: v1
kind: Service
metadata:
  name: mimock-svc
spec:
  type: NodePort
  selector:
    app: mimock
  ports:
    - protocol: TCP
      port: 8080
      targetPort: 8080
      nodePort: 30001
