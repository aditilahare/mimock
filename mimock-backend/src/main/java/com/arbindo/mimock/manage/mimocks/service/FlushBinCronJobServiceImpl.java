package com.arbindo.mimock.manage.mimocks.service;

import com.arbindo.mimock.entities.PlatformSettings;
import com.arbindo.mimock.manage.platformsettings.service.PlatformSettingsService;
import lombok.extern.log4j.Log4j2;
import org.apache.logging.log4j.Level;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;

@Service
@Log4j2
public class FlushBinCronJobServiceImpl implements FlushBinCronJobService {

    @Autowired
    PlatformSettingsService platformSettingsService;

    @Autowired
    DeleteMockService deleteMockService;

    @Override
    @Scheduled(cron = "${flush-bin-cron-expression}")
    public void flushBin() {
        PlatformSettings platformSettings = platformSettingsService.getDefaultPlatformSettings();
        if(!platformSettings.getIsFlushBinCronEnabled()){
            log.log(Level.INFO, "Flush Bin Cron Is Disabled. Skipping CRON scheduler job.");
            return;
        }
        log.log(Level.INFO, "Starting CRON scheduler job.");
        deleteMockService.flushDeletedMocks();
        log.log(Level.INFO, "Completed CRON scheduler job.");
    }
}
