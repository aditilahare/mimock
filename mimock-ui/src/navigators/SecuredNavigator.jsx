import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { Navigate } from 'react-router-dom';
import { useRecoilState } from 'recoil';
import { isTokenValid } from 'services/authentication/validateToken.service';
import showFullPageLoader from 'atoms/showFullPageLoader';

function Navigator({ children }) {
	const [, setOpenFullPageLoader] = useRecoilState(showFullPageLoader);
	const [isLoading, setIsLoading] = useState(true);
	const [isAuthTokenValid, setIsAuthTokenValid] = useState(false);

	useEffect(() => {
		setOpenFullPageLoader(true);

		isTokenValid()
			.then((status) => {
				setIsAuthTokenValid(status);
				setOpenFullPageLoader(false);
				setIsLoading(false);
			})
			.catch(() => {
				setIsAuthTokenValid(false);
				setOpenFullPageLoader(false);
				setIsLoading(false);
			});
	}, []);

	return (
		<If condition={!isLoading}>
			<Choose>
				<When condition={!isLoading && isAuthTokenValid}>{children}</When>
				<Otherwise>
					<Navigate to='/' replace={true} />
				</Otherwise>
			</Choose>
		</If>
	);
}

Navigator.propTypes = {
	children: PropTypes.node.isRequired,
};

export default Navigator;
