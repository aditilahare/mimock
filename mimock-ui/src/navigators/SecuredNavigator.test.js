import React from 'react';
import { render, waitFor, act } from '@testing-library/react';
import { isTokenValid } from 'services/authentication/validateToken.service';
import { Navigate } from 'react-router-dom';
import SecuredNavigator from './SecuredNavigator';

let mockedRecoilFn;
jest.mock('recoil', () => {
	mockedRecoilFn = jest.fn();
	return {
		atom: jest.fn(),
		useRecoilState: jest.fn(() => [true, mockedRecoilFn]),
	};
});
jest.mock('services/authentication/validateToken.service');
jest.mock('react-router-dom');

describe('SecuredNavigator', () => {
	const TestComponent = () => {
		return <div data-testid='child'>Child component</div>;
	};

	beforeAll(() => {
		Navigate.mockImplementation(() => {
			return <div data-testid='navigator'>Navigator</div>;
		});
	});

	it('should render child when token validation is successful', async () => {
		isTokenValid.mockResolvedValue(true);

		let tree;
		await act(async () => {
			tree = await render(
				<SecuredNavigator>
					<TestComponent />
				</SecuredNavigator>
			);
		});
		const { getByTestId, container, queryByTestId } = tree;

		await waitFor(() => {
			expect(mockedRecoilFn).toHaveBeenCalledTimes(2);
			expect(mockedRecoilFn).toHaveBeenNthCalledWith(1, true);
			expect(mockedRecoilFn).toHaveBeenNthCalledWith(2, false);

			expect(getByTestId('child')).toBeInTheDocument();
			expect(queryByTestId('fullpage-loader')).not.toBeInTheDocument();
			expect(queryByTestId('navigator')).not.toBeInTheDocument();
			expect(container).toMatchSnapshot();
		});
	});

	it('should navigate to home page when auth token validation fails', async () => {
		isTokenValid.mockRejectedValue(new Error('Validation failed'));

		let tree;
		await act(async () => {
			tree = await render(
				<SecuredNavigator>
					<TestComponent />
				</SecuredNavigator>
			);
		});
		const { getByTestId, container, queryByTestId } = tree;

		await waitFor(() => {
			expect(mockedRecoilFn).toHaveBeenCalledTimes(2);
			expect(mockedRecoilFn).toHaveBeenNthCalledWith(1, true);
			expect(mockedRecoilFn).toHaveBeenNthCalledWith(2, false);

			expect(getByTestId('navigator')).toBeInTheDocument();
			expect(queryByTestId('child')).not.toBeInTheDocument();
			expect(queryByTestId('fullpage-loader')).not.toBeInTheDocument();
			expect(container).toMatchSnapshot();
		});
	});
});
