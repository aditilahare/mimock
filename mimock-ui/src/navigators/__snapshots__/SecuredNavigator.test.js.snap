// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SecuredNavigator should navigate to home page when auth token validation fails 1`] = `
<div>
  <div
    data-testid="navigator"
  >
    Navigator
  </div>
</div>
`;

exports[`SecuredNavigator should render child when token validation is successful 1`] = `
<div>
  <div
    data-testid="child"
  >
    Child component
  </div>
</div>
`;
